### DISCUSSION TITLE
Dynamically set all properties of an extended class as not required

### DISCUSSION THREAD

**Discussion by EduardoFromTheFuture** [typegoose/typegoose#791] [Q&A]
In the code below, how I can have all the properties of EntityRequest as not required (required: false) without having to duplicate code?

I am extending the Entity model to define the EntityRequest model.
Both models should be the same, with the exception that all properties of EntityRequest should be not required.

The goal is to not repeat code. In other words, by dynamically making it so that every property of Entity will be considered in EntityRequest, without having to write code specifically for each of them. For example, the code of EntityRequest
```ts
@modelOptions({ schemaOptions: { timestamps: true } })
export class EntityRequest extends Entity {
  @prop({ required: false, unique: true })
  public legalEntityName!: string;

  @prop({ required: false, enum: ENUM_ENTITY_TYPE})
  public entityType!: string;
};
```

should be replaced by something like
```ts
@modelOptions({ schemaOptions: { timestamps: true } })
export class EntityRequest extends Entity {
  "for each property, set required: false"
};
```

---

```ts
import { prop, modelOptions, getModelForClass } from '@typegoose/typegoose';
import { ENUM_ENTITY_TYPE } from 'src/utils/enums'
import { countries } from 'src/utils/countriesList'
import { model } from 'mongoose'


@modelOptions({ schemaOptions: { timestamps: true } })
export class Entity {
  @prop({ required: true, unique: true })
  public legalEntityName!: string;

  @prop({ required: true, enum: ENUM_ENTITY_TYPE})
  public entityType!: string;

}

export const EntityModel = getModelForClass(Entity);


@modelOptions({ schemaOptions: { timestamps: true } })
export class EntityRequest extends Entity {
  @prop({ required: false, unique: true })
  public legalEntityName!: string;

  @prop({ required: false, enum: ENUM_ENTITY_TYPE})
  public entityType!: string;
};

export const EntityRequestModel = getModelForClass(EntityRequest);
```

**Comment by hasezoey** [comment]
no this is not supported by typegoose - you will have to "duplicate" code (also typegoose will likely never support such a thing)

if you really want to map properties, then you can try to [manually build the schema](https://typegoose.github.io/typegoose/docs/guides/advanced/manual-schema-modification) and add a mapper in-between (or use a plugin to map it)

**Comment by MikeKoval** [comment]
@PinkFromTheFuture i faced same request, wondering how did you resolve it? did you duplicate code?
i am not a fun of duplicating code also, trying to use base abstract class and duplicate definition of required fields in inherited one

**Reply by EduardoFromTheFuture** [reply]
Honestly I ended up migrating away from typegoose and mongoose all together and now I am back to relational databases while I call no-SQL databases "Glorified crap" =P

They have use-cases, but IMO, it should be very specific stuff, and not be used as basis for a whole platform.

**Reply by MikeKoval** [reply]
wondering what ORM do you use to work relational database? Prisma?

**Comment by EduardoFromTheFuture** [comment]
Yeah, Prisma.

But I believe that TypeORM would also be a solid choice.

On Fri, Jan 5, 2024 at 9:40 AM Mike Koval ***@***.***> wrote:

> wondering what ORM do you use to work relational database? Prisma?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/typegoose/typegoose/discussions/791#discussioncomment-8022811>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABSWIJIDQQCOGVK7H2QT6ITYM7YCDAVCNFSM6AAAAABBNAMS3KVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DAMRSHAYTC>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

<http://t-rize.io>

Eduardo Furtado

Chief Technology Officer

T-RIZE Group <http://t-rize.io>

***@***.***

+1 (800) 617 0366, Extension 704 <+18006170366,704>

2500-1155 René-Lévesque Blvd W,
Montréal, QC, Canada, H3B 3X7 <https://goo.gl/maps/HbJRqaRZ7EhkPimT8>

<https://www.linkedin.com/in/pinkfromthefuture/>

<https://github.com/PinkFromTheFuture>

Avis de confidentialité : Ce courriel et les pièces qui y sont jointes
contiennent de l'information confidentielle et peuvent être protégés par le
secret professionnel ou constituer de l’information privilégiée. Ils sont
destinés à l'usage exclusif de la (des) personne(s) à qui ils sont
adressés. Si vous n'êtes pas le destinataire visé ou la personne chargée de
transmettre ce document à son destinataire, vous êtes avisé par la présente
que toute divulgation, reproduction, copie, distribution ou autre
utilisation de cette information est strictement interdite. Si vous avez
reçu ce courriel par erreur, veuillez en aviser immédiatement l’expéditeur
par courriel ainsi que détruire et effacer l'information que vous avez
reçue de tout disque dur ou autre média sur lequel elle peut être
enregistrée et ne pas en conserver de copie. Merci de votre collaboration.


Notice of Confidentiality: This electronic mail message, including any
attachments, is confidential and may be privileged and protected by
professional secrecy. They are intended for the exclusive use of the
addressee. If you are not the intended addressee or the person responsible
for delivering this document to the intended addressee, you are hereby
advised that any disclosure, reproduction, copy, distribution or other use
of this information is strictly forbidden. If you have received this
document by mistake, please immediately inform the sender by email, destroy
and delete the information received from any hard disk or any media on
which it may have been registered and do not keep any copy. Thank you for
your cooperation.
